### PR TITLE
update kube-proxy to auto choose version

### PR DESCRIPTION
**Pull request by zjaco13**
*Issue #, if available:* #747

*Description of changes:*
Added version map for kube-proxy addon

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: lib/addons/kube-proxy/index.ts [hunk 1]
```diff
@@ -1,15 +1,54 @@
+import { CfnAddon, KubernetesVersion } from "aws-cdk-lib/aws-eks";
+import { Construct } from "constructs";
+import { ClusterInfo } from "../../spi";
+import { logger } from "../../utils";
 import { CoreAddOn } from "../core-addon";
 
+const versionMap: Map<KubernetesVersion, string> = new Map([
+    [KubernetesVersion.of("1.27"), "v1.27.1-eksbuild.1"],
```
Comment (elamaran11) [inline lib/addons/kube-proxy/index.ts:8]: Why is this line different?

**Review by elamaran11 — CHANGES_REQUESTED** [review]
@zjaco13 Have a minor comment. Also the doc is not updated.
@shapirov103 Is this approach good. CA uses helm approach and coreaddon props is readonly.

File: lib/addons/kube-proxy/index.ts (hunk 1 above)
Comment (zjaco13) [reply lib/addons/kube-proxy/index.ts:8]: The CDK team hasn't added Kubernetes version 1.27 to their enum list yet

File: lib/addons/kube-proxy/index.ts [hunk 2]
```diff
@@ -1,15 +1,54 @@
...
         super({
             addOnName: "kube-proxy",
-            version: version ?? "v1.25.6-eksbuild.1",
+            version: version ?? "auto",
             saName: "kube-proxy"
         });
+        this.version = version ?? "auto";
+    }
+
+    deploy(clusterInfo: ClusterInfo): Promise<Construct> {
```
Comment (shapirov103) [inline lib/addons/kube-proxy/index.ts:31]: Hey Zach, while this implementation works, it almost nullifies the inheritance  - there is very little reason to derive this from a core addon. 

I think we what we should do in this case is to enable this type of version selection for any core addon. 

That means that the core addon should be able to accept optionally something like a versionProviderFunction that takes cluster info and spits out the version. 

The kubeproxy will supply its function that will use the map. However, the default function will to just return whatever customers passed. 

In this way, other core addons, the number of which will grow, will be able to address this aspect without sacrificing the central implementation component of such addons.

**Review by shapirov103 — COMMENTED** [review]
@zjaco13 It is a great start, I like the map. Please see my comment on the design. Happy to discuss if needed.

File: lib/addons/kube-proxy/index.ts (hunk 2 above)
Comment (zjaco13) [reply lib/addons/kube-proxy/index.ts:31]: Yep, that approach makes a ton of sense.  We should discuss implementation.

File: lib/addons/core-addon/index.ts [hunk 3]
```diff
@@ -126,4 +139,15 @@ export class CoreAddOn implements ClusterAddOn {
         return result;
     }
 
-}
\ No newline at end of file
+    provideVersion(clusterInfo: ClusterInfo, versionMap: Map<KubernetesVersion, string>) : string {
+        let version: string;
+        if(!versionMap.get(clusterInfo.version)) {
```
Comment (shapirov103) [inline lib/addons/core-addon/index.ts:144]: nitpick: no need to get from the map twice on line 144 and 147. 
`let version = versionMap.get();`

**Review by shapirov103 — COMMENTED** [review]
All good, one small nit.

File: lib/addons/core-addon/index.ts (hunk 3 above)
Comment (zjaco13) [reply lib/addons/core-addon/index.ts]: resolved

**Review by shapirov103 — APPROVED** [review]
LGTM
pending e2e tests

**Comment by shapirov103** [comment]
@elamaran11 please review the latest

**Review by elamaran11 — APPROVED** [review]
LGTM

**Comment by elamaran11** [comment]
> @elamaran11 please review the latest

We can merge after e2e.

**Pull request merged**